docs/debugger_v2.md

```diff
@@ -0,0 +1,303 @@
+  program. Each alert indicates a certain anomaly that warrants attention. In
+  our case, this section highlights 442 NaN/∞ events with a salient pink-red
+  color. This confirms our suspicion that the model fails to learn because of
+  the presence of NaNs and/or infinities in its internal tensor values. We’ll
+  delve into these alerts shortly.
+- **Python Execution Timeline**: This is the upper half of the top-middle
+  section. It presents the full history of the eager execution of ops and
+  graphs. Each box of the timeline is marked by the initial letter of the op or
+  graph’s name (e.g., “T” for the “TensorSliceDataset” op, “m” for the “model”
+  `tf.function`). We can navigate this timeline by using the navigation buttons
+  and the scrollbar above the timeline.
+- **Graph Execution** : Located at the top-right corner of the GUI, this section
+  will be central to our debugging task. It contains a history of all the
+  floating-dtype tensors computed inside graphs (i.e., compiled by
+  `@tf-function`s).
+- **Graph Structure** (bottom half of the top-middle section), **Source Code**
+  (bottom-left section), and **Stack Trace** (bottom-right section) are
+  initially empty. Their contents will be populated when we interact with the
+  GUI. These three sections will also play important roles in our debugging
+  task.
+
+Having oriented ourselves to the organization of the UI, let’s take the
+following steps to get to the bottom of why the NaNs appeared. First, click the
+**NaN/∞** alert in the Alerts section. This automatically scrolls the list of
+600 graph tensors in the Graph Execution section and focuses on the #67, which
+is a tensor named “Log:0” generated by a `Log` (natural logarithm) op. A salient
+red color that highlights a -∞ element among the 1000 elements of the 2D float32
+tensor. This is the first tensor in the TF2 program’s runtime history that
+contained any NaN or infinity: tensors computed before it do not contain NaN or
+∞; many (in fact, most) tensors computed afterwards contain NaNs. We can confirm
+this by scrolling up and down the Graph Execution list. This observation
+provides a strong hint that the `Log` op beingthe source of the numerical
+instability in this TF2 program.
+
+![Debugger V2: Nan / Infinity alerts and graph execution list](./images/debugger_v2_2_nan_inf_alerts.png)
+
+Why does this `Log` op spit out a -∞? Answering that question requires examining
+the input to the op. Clicking on the name of the tensor (“Log:0”) brings up a
+simple but informative visualization of the `Log` op’s vicinity in its
+TensorFlow graph in the Graph Structure section. Note the top-to-bottom
+direction of information flow. The op itself is shown in the bold in the middle.
+Immediately above it, we can see a Placeholder op provides the one and only input
+to the `Log` op. Where is the tensor generated by this logits Placeholder in the
+Graph Execution list? By using the background color and the “input” labels as
```

Comment:
background color?  They all look grey in the images.

Reply:
Ah - I was anticipating an in-flight PR (#3765) when writing this. The PR has now landed and I've updated the screenshot here. Thanks for the catch.